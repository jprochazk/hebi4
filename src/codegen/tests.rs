use std::fs::read_to_string;

use super::emit;
use crate::{parser::parse, token::tokenize};

uitest::uitest!("tests/inputs/codegen/*.hi", "codegen", |path| {
    let input = read_to_string(path).unwrap();
    let tokens = tokenize(&input);
    let ast = match parse(&tokens) {
        Ok(ast) => ast,
        Err(err) => {
            panic!("{}", err.render(&input));
        }
    };
    let (snapshot, failure) = match emit(&ast) {
        Ok(chunk) => (format!("OK\n{}", chunk.disasm(&input)), false),
        Err(err) => (format!("ERROR\n{}", err.render(&input)), true),
    };

    #[cfg(not(miri))]
    {
        insta::assert_snapshot!(snapshot);
        let _ = failure;
    }

    #[cfg(miri)]
    {
        if failure {
            panic!("{snapshot}");
        } else {
            eprintln!("{snapshot}");
        }
    }
});
