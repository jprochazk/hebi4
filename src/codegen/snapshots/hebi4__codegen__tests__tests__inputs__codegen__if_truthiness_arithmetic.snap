---
source: src/codegen/tests.rs
expression: snapshot
---
OK
.fn @main
  params 0
  stack 5
  .locals
    a = r1
    b = r2
  .code
   0  lsmi r1, 5
   1  lsmi r2, 3
   2  add r3, r1, r2
   3  istrue r3
   4  jmp 3   ; to 7
   5  lsmi r0, 1
   6  ret
   7  sub r3, r1, 5
   8  istrue r3
   9  jmp 3   ; to 12
  10  lsmi r0, 2
  11  ret
  12  mul r0, r1, 0
  13  istrue r0
  14  jmp 4   ; to 18
  15  lsmi r0, 3
  16  ret
  17  lnil r0
  18  ret
